exec/command.go

```diff
@@ -0,0 +1,186 @@
+// as Path.
+// By default a command is run in the current working directory.
+func Command(name string, arg ...string) *Cmd {
+	return &Cmd{
+		path:         name,
+		args:         arg,
+		dir:          ".",
+		debugfFn:     DefaultDebugfFn,
+		debugfPrefix: DefaultDebugPrefix,
+	}
+}
+
+// ShellCommand executes a command in sh shell.
+func ShellCommand(cmd string) *Cmd {
+	return Command("sh", "-c", cmd)
+}
+
+// Directory changes the directory in which the command is executed.
+func (c *Cmd) Directory(dir string) *Cmd {
+	c.dir = dir
+	return c
+}
+
+// DebugfFunc sets the debug function for the command. It accepts a
+// printf-style printf function and call it for every line that the command
+// prints to STDOUT and STDERR when it's run.
+func (c *Cmd) DebugfFunc(fn func(format string, arg ...interface{})) *Cmd {
+	c.debugfFn = fn
+	return c
+}
+
+// DebugfPrefix sets a prefix that is prepended to the message that is passed to the Debugf function.
+func (c *Cmd) DebugfPrefix(prefix string) *Cmd {
+	c.debugfPrefix = prefix
+	return c
+}
+
+// ExpectSuccess if called, Run() will return an error if the command did not
+// exit with code 0.
+func (c *Cmd) ExpectSuccess() *Cmd {
+	c.expectSuccess = true
+	return c
+}
+
+func cmdString(cmd *exec.Cmd) string {
+	// cmd.Args[0] contains the command name, cmd.Path the absolute command path,
+	// omit cmd.Args[0] from the string
+	if len(cmd.Args) > 1 {
+		return fmt.Sprintf("%s %v", cmd.Path, strings.Join(cmd.Args[1:], " "))
+	}
+
+	return cmd.Path
+}
+
+// Result describes the result of a run Cmd.
+type Result struct {
+	Command  string
+	Dir      string
+	Output   []byte
+	ExitCode int
+}
+
+// StrOutput returns Result.Output as string.
+func (r *Result) StrOutput() string {
+	return string(r.Output)
+}
+
+// ExpectSuccess if the ExitCode in Result is not 0, the function returns an
+// ExitCodeError for the execution.
+func (r *Result) ExpectSuccess() error {
+	if r.ExitCode == 0 {
+		return nil
+	}
+
+	return ExitCodeError{Result: r}
+}
+
+// Run executes the command.
+func (c *Cmd) Run() (*Result, error) {
+	cmd := exec.Command(c.path, c.args...)
+	cmd.Dir = c.dir
+
+	outReader, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+	cmd.Stderr = cmd.Stdout
+
+	c.debugfFn(c.debugfPrefix+"running '%s' in directory '%s'", cmdString(cmd), cmd.Dir)
+	err = cmd.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	var outBuf bytes.Buffer
+	firstline := true
+	in := bufio.NewScanner(outReader)
+	for in.Scan() {
+		if firstline {
+			firstline = false
+		} else {
+			outBuf.WriteRune('\n')
+		}
+
+		c.debugfFn(c.debugfPrefix + in.Text())
+
+		outBuf.Write(in.Bytes())
+	}
+
+	if err := in.Err(); err != nil {
+		cmd.Wait()
+
+		return nil, err
+	}
+
+	var exitCode int
+	err = cmd.Wait()
+	if err != nil {
+		if ee, ok := err.(*exec.ExitError); ok {
+			if status, ok := ee.Sys().(syscall.WaitStatus); ok {
+				exitCode = status.ExitStatus()
+			}
```

Comment:
I feel like there should be an *else* statement here

Reply:
```
else {
    exitCode = 0
}
```
?
I think it only adds unnecessary lines. :-)

Reply:
We have three options:
1. The shortest code: `exitCode = ee.Sys().(syscall.WaitStatus).ExitStatus()`
2. Test type and ignore if type is not syscall.WaitStatus
3. Test type and return error if type don't match

Option 2 will silence a type conversion failure and assume that exitCode is 0. In my opinion the other two options are better.

Reply:
thanks a lot!
I fixed it like discussed verbally